### ISSUE TITLE
LLDB build for MinGW broken by unexpected include of unistd.h

### ISSUE THREAD

**Issue by mstorsjo** [llvm/llvm-project#76050]
Builds of LLDB for MinGW targets, with libc++ as the C++ standard library, was recently broken by an entirely unrelated change in libc++, commit 2fd4084fca0c474c2232533263182945bd1ebe17. An example of such a failed build: https://github.com/mstorsjo/llvm-mingw/actions/runs/7269329294/job/19809313900

To lay out the situation; mingw-w64's `getopt.h` header, https://github.com/mingw-w64/mingw-w64/blob/v11.0.1/mingw-w64-headers/crt/getopt.h, does define `optreset` only if `_BSD_SOURCE` was defined.

LLDB's use of `getopt.h` does require using `optreset`. To make sure this is available, LLDB does define `_BSD_SOURCE` right before including `getopt.h`: https://github.com/llvm/llvm-project/blob/main/lldb/include/lldb/Host/HostGetOpt.h#L12-L18

Mingw-w64's `unistd.h` also includes `getopt.h`, as one can expect to have `getopt` available after including `unistd.h`: https://github.com/mingw-w64/mingw-w64/blob/v11.0.1/mingw-w64-headers/crt/unistd.h#L12

Now at this point, the innocent libc++ change in 2fd4084fca0c474c2232533263182945bd1ebe17 caused more of libc++'s headers to include `<unistd.h>`, e.g. here: https://github.com/llvm/llvm-project/blob/2fd4084fca0c474c2232533263182945bd1ebe17/libcxx/include/print#L52-L54 (that line was preexisting, but the change made more headers include `<print>`).

Thus, LLDB includes a bunch of C++ standard headers, which end up bringing in `unistd.h`, which includes `getopt.h`. This then renders a later `#define _BSD_SOURCE` `#include <getopt.h>` in `HostGetOpt.h` without effect.

I guess this generally is a brittle pattern - defines such as `_BSD_SOURCE` or `_GNU_SOURCE` generally need to be defined on the command line, to be sure they applies as headers can be included transitively, earlier than expected.

Alternatively, the mingw-w64 `getopt.h` header could be adapted to react to being reincluded with `_BSD_SOURCE` defined, after first being included once without it.

CC @DavidSpickett

**Comment by DavidSpickett** [comment]
Seems that moving the define up into cmake config is the accepted way to fix it. Judging by other references to `_GNU_SOURCE` and `_BSD_SOURCE`.

I've not used these macros much before, so I assume defining `_BSD_SOURCE` for potentially many more includes is ok. As there are places in llvm already doing this for `_GNU_SOURCE` when building against glibc.

**Comment by mstorsjo** [comment]
FWIW, I somehow didn't get any mail notification for this comment... I went ahead and pushed the mingw-w64 header fix in https://github.com/mingw-w64/mingw-w64/commit/940e8faa948f4fcb35b6a724d48f706193e07c85.

> Seems that moving the define up into cmake config is the accepted way to fix it. Judging by other references to `_GNU_SOURCE` and `_BSD_SOURCE`.

Yep; I did test setting `_BSD_SOURCE` at the same place in `lldb/cmake/modules/LLDBGenerateConfig.cmake` as where we're currently setting `_GNU_SOURCE`, but that didn't seem to have any effect when building the failing source file. So it would need to be set somewhere else...

> I've not used these macros much before, so I assume defining `_BSD_SOURCE` for potentially many more includes is ok. As there are places in llvm already doing this for `_GNU_SOURCE` when building against glibc.

Yes, I guess it should be ok in general. Worst case, one would have conflicts if user code would be using other symbols that happen to clash with symbols taken by bsd specific APIs, but as a cross platform codebase, that's quite unlikely.

> And as this is needed for `_WIN32` only

Actually, it's for `_WIN32` and not `_MSC_VER`, i.e. not MSVC but only mingw.

> you wouldn't have `_BSD_SOURCE` and `_GNU_SOURCE` defined at the same time, right? Maybe that's not an issue.

Well not necessarily (also remember mingw is considered "windows-gnu" in some contexts); the mingw headers also do contain a handful of GNU extensions that are hidden unless `_GNU_SOURCE` is defined. And I don't think those two are considered mutually exclusive either, it's just whether to opt in to extensions from either vendor family.
